### PR TITLE
Annotate System.Data.{Odbc,OleDb} for nullability

### PR DESCRIPTION
**Pull request by roji**
* As before, no behavior changes were done, only annotations (but see comments below)
* Nullability warnings are disabled for OleDb, see https://github.com/roji/runtime/tree/SystemDataOleDbBuildHack for a csproj hack to make these appear
* Unfortunately there don't seem to be any warnings about discrepancies between implementation (src) and interface (ref). I did the best I could to keep them in sync, but it would be better to have automatic verification.
* Also marked System.Data.DataSetExtensions as annotated - it's a facade with no actual types.

/cc @jeffhandley @ajcvickers

### CODE REVIEW

File: src/libraries/System.Data.OleDb/src/OleDbCommand.cs [hunk 1]
```diff
@@ -1275,17 +1275,16 @@ private void ProcessResultsNoReset(OleDbHResult hr)
...
             }
             Debug.Assert(!_isPrepared, "null command isPrepared");
 
-            OleDbConnection? connection = _connection;
-            if (null == connection)
-            {
-                connection.CheckStateOpen(ODB.Properties);
-            }
+            OleDbConnection connection = _connection;
+            connection.CheckStateOpen(ODB.Properties);
```
Comment (roji) [inline src/libraries/System.Data.OleDb/src/OleDbCommand.cs:1287]: Tiny behavioral change here

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs [hunk 2]
```diff
@@ -407,7 +407,10 @@ public override string GetDataTypeName(int i)
                 {
                     info._typename = GetColAttributeStr(i, ODBC32.SQL_DESC.TYPE_NAME, ODBC32.SQL_COLUMN.TYPE_NAME, ODBC32.HANDLER.THROW);
                 }
+// TODO: Behavior change probably needed here - when there's no data type, we mostly probably need to throw instead of returning null.
```
Comment (roji) [inline src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs:410]: Scoping this out, will submit a separate PR soon.

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs [hunk 3]
```diff
@@ -2250,7 +2253,10 @@ public override DataTable GetSchemaTable()
             }
             if (_noMoreResults)
             {
+// TODO: Behavior change (https://github.com/dotnet/runtime/issues/509)
+#nullable disable
```
Comment (roji) [inline src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs:2257]: Scoping this out, will submit a separate PR soon.

File: src/libraries/System.Data.OleDb/src/OleDbDataReader.cs [hunk 4]
```diff
@@ -273,7 +273,10 @@ public override DataTable GetSchemaTable()
                 //GetSchemaTable() is defined to return null after NextResult returns false
                 //throw ADP.DataReaderNoData();
             }
+// TODO: Behavior change (https://github.com/dotnet/runtime/issues/509)
+#nullable disable
```
Comment (roji) [inline src/libraries/System.Data.OleDb/src/OleDbDataReader.cs:277]: Scoping this out, will submit a separate PR soon.

**Comment by roji** [comment]
@stephentoub @jeffhandley the build failures seem to be because System.Data.Odbc is being built for older TFMs (netfx, netstandard2.0) where nothing is annotated, so Debug.Assert isn't recognized (when building locally, presumably for .NET Core, there are no errors/warnings). Nullability warnings/errors should probably be disabled for older TFMs here, right?

**Comment by jeffhandley** [comment]
> Nullability warnings/errors should probably be disabled for older TFMs here, right?

I'm not sure myself.  @ericstj, do you know?

**Review by jeffhandley — COMMENTED** [review]
I'm at 37/200 files viewed. So far so good. And thanks for fixing the spelling errors in comments!

File: src/libraries/Common/src/System/Data/Common/AdapterUtil.cs [hunk 5]
```diff
@@ -22,11 +19,11 @@ internal static partial class ADP
     {
         // NOTE: Initializing a Task in SQL CLR requires the "UNSAFE" permission set (https://docs.microsoft.com/en-us/dotnet/framework/performance/sql-server-programming-and-host-protection-attributes)
         // Therefore we are lazily initializing these Tasks to avoid forcing customers to use the "UNSAFE" set when they are actually using no Async features
-        private static Task<bool> _trueTask;
-        internal static Task<bool> TrueTask => _trueTask ?? (_trueTask = Task.FromResult(true));
+        private static Task<bool>? _trueTask;
+        internal static Task<bool> TrueTask => _trueTask ??= Task.FromResult(true);
```
[the remark this answers is not in the document]
Comment (roji) [reply src/libraries/Common/src/System/Data/Common/AdapterUtil.cs:23]: I actually had a hard time resisting a much more general cleanup - there's so much C# goodness that has been added since this code has been written. But these annotation PRs are already huge, so I think it's best for everyone to really stick to annotating. We can always do some cleanup (even automating) after this).

**Comment by ericstj** [comment]
>> Nullability warnings/errors should probably be disabled for older TFMs here, right?

> I'm not sure myself. @ericstj, do you know

Looks to me like folks are suppressing the warnings on other frameworks: https://github.com/dotnet/runtime/blob/3f7202ba9f2f9a604816530f9d3154b0ee279a21/src/libraries/Directory.Build.targets#L9-L10

File: src/libraries/Common/src/System/Data/Common/AdapterUtil.cs [hunk 6]
```diff
@@ -480,15 +477,15 @@ internal static Exception InvalidDataLength(long length)
 
         internal static int DstCompare(string strA, string strB) => CultureInfo.CurrentCulture.CompareInfo.Compare(strA, strB, ADP.DefaultCompareOptions);
 
-        internal static bool IsEmptyArray(string[] array) => (null == array) || (0 == array.Length);
+        internal static bool IsEmptyArray([NotNullWhen(false)] string?[]? array) => (null == array) || (0 == array.Length);
```
Comment (krwq) [inline src/libraries/Common/src/System/Data/Common/AdapterUtil.cs:480]: I think `NotNullWhen` is only for `out` (possibly `ref`) args

File: src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs [hunk 7]
```diff
@@ -515,20 +515,20 @@ internal static string[] RemoveDuplicates(string[] restrictions)
                     count = 0;
                     for (int i = 0; i < restrictions.Length; ++i)
                     {
-                        if (null != restrictions[i])
+                        if (restrictions[i] is { } restriction)
```
Comment (krwq) [inline src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: what does `is { }` mean?

Comment (roji) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: This is just a way to use pattern making to check if a value isn't null

Comment (stephentoub) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: `!= null` or `is not null` would be better.

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnectionOpen.cs [hunk 8]
```diff
@@ -32,7 +32,9 @@ public override string ServerVersion
         {
             get
             {
-                return OuterConnection.Open_GetServerVersion();
+                // TODO: This seems like it returns null if the connection is open, whereas the docs say it should throw
```
Comment (krwq) [inline src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnectionOpen.cs:35]:
```suggestion
                // TODO-NULLABLE: This seems like it returns null if the connection is open, whereas the docs say it should throw
```

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs (hunk 2 above)
Comment (krwq) [inline src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs:410]:
```suggestion
// TODO-NULLABLE: Behavior change probably needed here - when there's no data type, we mostly probably need to throw instead of returning null.
```

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs [hunk 9]
```diff
@@ -2250,7 +2253,10 @@ public override DataTable GetSchemaTable()
             }
             if (_noMoreResults)
             {
+// TODO: Behavior change (https://github.com/dotnet/runtime/issues/509)
```
Comment (krwq) [inline src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs:2256]:
```suggestion
// TODO-NULLABLE: Behavior change (https://github.com/dotnet/runtime/issues/509)
```

File: src/libraries/System.Data.OleDb/ref/System.Data.OleDb.cs [hunk 10]
```diff
@@ -180,9 +182,9 @@ public sealed partial class OleDbDataReader : System.Data.Common.DbDataReader
         public override void Close() { }
         public override bool GetBoolean(int ordinal) { throw null; }
         public override byte GetByte(int ordinal) { throw null; }
-        public override long GetBytes(int ordinal, long dataIndex, byte[] buffer, int bufferIndex, int length) { throw null; }
+        public override long GetBytes(int ordinal, long dataIndex, byte[]? buffer, int bufferIndex, int length) { throw null; }
         public override char GetChar(int ordinal) { throw null; }
-        public override long GetChars(int ordinal, long dataIndex, char[] buffer, int bufferIndex, int length) { throw null; }
+        public override long GetChars(int ordinal, long dataIndex, char[]? buffer, int bufferIndex, int length) { throw null; }
```
Comment (krwq) [inline src/libraries/System.Data.OleDb/ref/System.Data.OleDb.cs:187]: is null buffer actually allowed or is it only for zero length?

File: src/libraries/System.Data.OleDb/src/ColumnBinding.cs [hunk 11]
```diff
@@ -948,7 +948,8 @@ internal OleDbDataReader Value_HCHAPTER()
             Debug.Assert(NativeDBType.HCHAPTER == DbType, "Value_HCHAPTER");
             Debug.Assert(DBStatus.S_OK == StatusValue(), "Value_HCHAPTER");
 
-            return DataReader().ResetChapter(IndexForAccessor, IndexWithinAccessor, RowBinding, ValueOffset);
+            // TODO: This shouldn't return null
```
Comment (krwq) [inline src/libraries/System.Data.OleDb/src/ColumnBinding.cs:951]:
```suggestion
            // TODO-NULLABLE: This shouldn't return null
```

**Comment by krwq** [comment]
I'm at 65/120 and have not seen anything blocking so far - my eyes got a bit blurry with question marks flying so will likely not finish today.

The only comment is to replace all `TODO` with `TODO-NULLABLE` for easier tracking

File: src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs (hunk 7 above)
Comment (roji) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: `!= null` is the check being replaced, the problem is that since we're checking 

`is not null` is C# 9, right? Is it OK to use this already in runtime?

File: src/libraries/System.Data.OleDb/ref/System.Data.OleDb.cs (hunk 10 above)
Comment (roji) [reply src/libraries/System.Data.OleDb/ref/System.Data.OleDb.cs:187]: A null buffer makes the method just return the number of characters in the column, without returning it in any way ([this is even in the docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqldatareader.getchars?view=sqlclient-dotnet-core-2.0#Microsoft_Data_SqlClient_SqlDataReader_GetChars_System_Int32_System_Int64_System_Char___System_Int32_System_Int32_)...).

**Comment by roji** [comment]
@ericstj (or anyone else) any idea about the nullability errors in the build? It's odd that they're only in that one config, and I don't get them when building locally...,

**Comment by stephentoub** [comment]
> (or anyone else) any idea about the nullability errors in the build? It's odd that they're only in that one config, and I don't get them when building locally...

The reason we suppress nullability warnings in https://github.com/dotnet/runtime/blob/3f7202ba9f2f9a604816530f9d3154b0ee279a21/src/libraries/Directory.Build.targets#L9-L10 for downlevel targets is because those surface areas don't have nullable annotations / attributes.  That includes `Debug.Assert`, which means if you have code like `Debug.Assert(foo != null); foo.Blah();`, without the `[DoesNotReturnIf(false)]` attribute on the assert, the compiler doesn't know that `foo.Blah()` is valid and warns that `foo` might be null.  So, the most likely cause here is this library is building against a target that is not yet included in that Directory.Build.targets exclusion, and the fix would be to extend those conditions to make sure it's included.

**Comment by ericstj** [comment]
Are you doing an `-allconfigurations` build locally?  The other legs are all for our vertical configurations where we execute tests, whereas `allconfigurations` includes other configurations which ship in the package.  From the looks of the [log](https://dev.azure.com/dnceng/_apis/resources/Containers/4656787/Windows_NT_Libraries%20Build%20Windows_NT%20allConfigurations%20x64%20Debug?itemPath=Windows_NT_Libraries%20Build%20Windows_NT%20allConfigurations%20x64%20Debug%2FBuild.binlog) the framework you're failing on is 'netcoreapp2.0-FreeBSD`.

We should probably just change the condition on nowarn to suppress on `'$(TargetFrameworkIdentifier)' == '.NETCoreApp` and $([MSBuild]::VersionLessThan($(TargetFrameworkVersion), '3.0')` instead of only `netcoreapp2.1` as it does today.

**Comment by ericstj** [comment]
I went ahead and pushed a commit to suppress nullable warnings on all old netcoreapp frameworks as I suggested.

**Comment by roji** [comment]
Thanks @ericstj!

**Comment by ericstj** [comment]
> Thanks @ericstj!

Sure thing!  @krwq @jeffhandley I think this is still pending review of annotations.

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcParameterHelper.cs [hunk 12]
```diff
@@ -233,29 +234,19 @@ private byte ValueScaleCore(object value)
...
-                {
-                    return 1;
-                }
+                    string svalue => svalue.Length,
+                    byte[] bvalue => bvalue.Length,
+                    char[] cvalue => cvalue.Length,
+                    byte _ => 1,
+                    char _ => 1,
+                    _ => 0
+                };
```
Comment (jeffhandley) [inline src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcParameterHelper.cs:249]: 👍

File: src/libraries/System.Data.OleDb/src/OleDbCommand.cs [hunk 13]
```diff
@@ -802,19 +804,20 @@ private int ExecuteCommand(CommandBehavior behavior, out object executeResult)
...
-            Bindings bindings = ParameterBindings;
+            tagDBPARAMS? dbParams = null;
+            RowBinding? rowbinding = null;
+            Bindings? bindings = ParameterBindings;
             bool mustRelease = false;
 
             RuntimeHelpers.PrepareConstrainedRegions();
             try
             {
+                // TODO: Code below seems to assume that bindings will always be non-null
```
Comment (jeffhandley) [inline src/libraries/System.Data.OleDb/src/OleDbCommand.cs:815]: Nit: Did you want to use `TODO-NULLABLE` here too?

File: src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs [hunk 14]
```diff
@@ -410,26 +411,26 @@ private string GetParameterName(string neededCollectionName, int neededRestricti
...
-            exactCollectionName = requestedCollectionRow[collectionNameColumn, DataRowVersion.Current] as string;
+            exactCollectionName = (requestedCollectionRow[collectionNameColumn, DataRowVersion.Current] as string)!;
 
             if (ADP.IsEmptyArray(restrictions) == false)
             {
-                for (int i = 0; i < restrictions.Length; i++)
+                for (int i = 0; i < restrictions!.Length; i++)
                 {
-                    if ((restrictions[i] != null) && (restrictions[i].Length > 4096))
+                    if ((restrictions?.Length > 4096))
```
Comment (jeffhandley) [inline src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs:433]: It's the little things in life. This is one of the few places through this PR where the nullable reference types feature actually _reduces noise_ in the code. 😃

**Review by jeffhandley — COMMENTED** [review]
I'm at 80/121 files (although I started moving bottom up here recently). Will keep going tomorrow.

File: src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs (hunk 14 above)
Comment (ericstj) [reply src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs:433]: Was it intentional to check length of array rather than element?

Comment (jeffhandley) [reply src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs:433]: Funny that I missed that detail after being enamored at the collapsing of the code. 🙃

File: src/libraries/System.Data.OleDb/src/OleDbConnectionFactory.cs [hunk 15]
```diff
@@ -31,13 +31,14 @@ public override DbProviderFactory ProviderFactory
             }
         }
 
-        protected override DbConnectionInternal CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningObject)
+        protected override DbConnectionInternal CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, object poolGroupProviderInfo, DbConnectionPool? pool, DbConnection? owningObject)
         {
-            DbConnectionInternal result = new OleDbConnectionInternal((OleDbConnectionString)options, (OleDbConnection)owningObject);
+            // TODO: owningObject may actually be null (see DbConnectionPool.CreateObject), in which case this will throw...
```
Comment (jeffhandley) [inline src/libraries/System.Data.OleDb/src/OleDbConnectionFactory.cs:36]: Not: TODO-NULLABLE here?

File: src/libraries/System.Data.OleDb/src/OleDbConnectionInternal.cs [hunk 16]
```diff
@@ -380,11 +380,12 @@ internal object GetDataSourcePropertyValue(Guid propertySet, int propertyID)
             return dbprops[0].dwStatus;
         }
 
-        internal DataTable BuildInfoLiterals()
+        internal DataTable? BuildInfoLiterals()
         {
             using (IDBInfoWrapper wrapper = IDBInfo())
             {
                 UnsafeNativeMethods.IDBInfo dbInfo = wrapper.Value;
+                // TODO: check may not be necessary (and thus method may return non-nullable)
```
Comment (jeffhandley) [inline src/libraries/System.Data.OleDb/src/OleDbConnectionInternal.cs:388]: Nit: TODO-NULLABLE here?

**Review by jeffhandley — COMMENTED** [review]
A bit more progress reviewing but still not done.

File: src/libraries/System.Data.OleDb/src/OleDbDataReader.cs (hunk 4 above)
Comment (jeffhandley) [reply src/libraries/System.Data.OleDb/src/OleDbDataReader.cs:277]: Did you want to use `TODO-NULLABLE` here too?

File: src/libraries/System.Data.OleDb/src/OleDbDataReader.cs [hunk 17]
```diff
@@ -1021,7 +1025,8 @@ public override Type GetFieldType(int index)
         {
             if (null != _metadata)
             {
-                return _metadata[index].type.dataType;
+                // TODO: Should throw if null (empty), though it probably doesn't happen
```
Comment (jeffhandley) [inline src/libraries/System.Data.OleDb/src/OleDbDataReader.cs:1028]:
```suggestion
                // TODO-NULLABLE: Should throw if null (empty), though it probably doesn't happen
```

**Review by jeffhandley — CHANGES_REQUESTED** [review]
Finished reviewed.  There were several places where you might want to use `TODO-NULLABLE` instead of `TODO` (following the pattern seen in other places in this PR).  Just suggestions though.

I think what @ericstj spotted [in `DbMetaDataFactory.cs`](https://github.com/dotnet/runtime/pull/39597/files/ada6f2accf4e0b92785f19ebb1fb9b360d895f4f#r459828812) is a lurking regression.  I don't know if there's existing test coverage that should have caught it though.

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs [hunk 18]
```diff
@@ -499,7 +500,7 @@ private string GetInfoStringUnhandled(ODBC32.SQL_INFO info, bool handleError)
                 }
                 else if (handleError)
                 {
-                    this.HandleError(ConnectionHandle, retcode);
+                    this.HandleError(connectionHandle, retcode);
```
[the remark this answers is not in the document]
Comment (roji) [reply src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs:503]: Yeah, no functional difference - I changed this to silence the nullability warning, as the nullability check above is on connectionHandle - which was just copied from ConnectionHandle - and neither was changed in the code between. So barring concurrency (which I think is not an issue here) the two are equivalent.

If you have any doubts I can change this back and add a bang though.

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnectionHelper.cs [hunk 19]
```diff
@@ -152,9 +152,8 @@ internal void AddWeakReference(object value, int tag)
 
         protected override DbCommand CreateDbCommand()
         {
-            DbCommand command = null;
             DbProviderFactory providerFactory = ConnectionFactory.ProviderFactory;
-            command = providerFactory.CreateCommand();
+            var command = providerFactory.CreateCommand()!;
```
[the remark this answers is not in the document]
Comment (roji) [reply src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnectionHelper.cs:156]: Good point, changed.

File: src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs (hunk 14 above)
Comment (roji) [reply src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbMetaDataFactory.cs:433]: Thanks for catching this @ericstj...! Correcting.

File: src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (jeffhandley) [reply src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs:503]: Cool--thanks for verifying. What you have looks more correct, I just wanted to confirm.

**Comment by roji** [comment]
@jeffhandley @krwq and the others, thanks for the thorough reviewing, I've integrated all your comments.

File: src/libraries/System.Data.OleDb/src/OleDbDataReader.cs [hunk 20]
```diff
@@ -2257,21 +2262,22 @@ private void AppendSchemaUniqueIndexAsKey(Hashtable baseColumnNames, object[] re
...
         {
             List<MetaData> metainfo = new List<MetaData>();
 
-            object hiddenColumns = null;
+            object? hiddenColumns = null;
             using (OleDbDataReader dataReader = new OleDbDataReader(_connection, _command, int.MinValue, 0))
             {
                 dataReader.InitializeIRowset(rowset, ChapterHandle.DB_NULL_HCHAPTER, IntPtr.Zero);
-                dataReader.BuildSchemaTableInfo(rowset, true, false);
+                // TODO: BuildSchemaTableInfo asserts that rowset isn't null, but doesn't do anything with it
```
Comment (jeffhandley) [inline src/libraries/System.Data.OleDb/src/OleDbDataReader.cs:2279]: Nit: another `TODO` that could be `TODO-NULLABLE` if you want.

```suggestion
                // TODO-NULLABLE: BuildSchemaTableInfo asserts that rowset isn't null, but doesn't do anything with it
```

**Review by jeffhandley — APPROVED** [review]
Thank you for all of the effort you put into this!!

File: src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs (hunk 7 above)
Comment (roji) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: FYI am going to merge as-is, but we can definitely replace these with the nicer `is not null` once we can use C# 9 here.

Comment (stephentoub) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: What's the benefit of this complicated syntax over != null here?

Comment (stephentoub) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: > is not null is C# 9, right? Is it OK to use this already in runtime

Yes and yes

Comment (jeffhandley) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: There's a tiny benefit of capturing `restriction`, that will be realized more when using `is not null`.

Comment (stephentoub) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: Sure, but neither `{ }` nor `is not null` are necessary to achieve that:
```C#
if (restrictions[i] is string restriction)
{
    ...
}
```

Comment (jeffhandley) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: Ah, yeah that'd be better indeed.

Comment (roji) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs:518]: The slight disadvantage of `restrictions[i] is string restriction` is that it's ambiguous with regards to whether we're checking that something isn't null, or that it's a string (as opposed to something else). FWIW Resharper/Rider warn about this.

I tried `restrictions[i] is not null restriction`, assuming that's how the C# 9 feature looks like, but it doesn't seem to compile. I'll just switch to `restrictions[i] is string restriction` for now as suggested.

**Review by jeffhandley — APPROVED** [review]

File: src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs (hunk 7 above)
Comment (stephentoub) [reply src/libraries/System.Data.Odbc/src/Common/System/Data/Common/DBConnectionString.cs]: If you see it as a disadvantage, then just do the super simple thing:
```C#
foreach (string? restriction in restrictions)
{
    if (restriction != null)
    {
        tmp[count++] = restriction;
    }
}
```
or
```C#
for (int i = 0; i < restrictions.Length; i++)
{
    string? restriction = restrictions[i];
    if (restriction != null)
    {
        tmp[count++] = restriction;
    }
}
```
or just suppress the warning:
```C#
for (int i = 0; i < restrictions.Length; i++)
{
    if (restrictions[i] != null)
    {
        tmp[count++] = restrictions[i]!; // TODO-NULLABLE: Indexer nullability tracked (https://github.com/dotnet/roslyn/issues/34644)
    }
}
```
as we do in dozens of other places in dotnet/runtime... the more we work around it, the less likely it'll be the compiler "fixes" it.
cc: @cston

Not everything needs to use fancy syntax.
